Improve max/min to handle NaN, -0.0f/-0.0d correctly

saifk16 opened the pull request:
In this PR, I made some changes to the `jpf-symbc/src/classes/java/lang/Math.java`:

**Handles**

- For double and float Math.min() and Math.max()
- NaN (Not a Number) [Double.NaN/Float.NaN]
- 0.0f and -0.0f / 0.0d and -0.0d

 ```java
     if(a != a) // NaN is not equal to itself
        return a;
  ```
  
  ```java
      return (1.0f / a > 0) ? a : b; 
  ```
We can also use raw conversion of bits as Nan cannot map to -0.0.
@yannicnoller @corinus @sohah

saifk16 committed:
Handle NaN and -0.0f/-0.0d max/min

sohah commented:
@saifk16 , thanks for your contribution. This pull request is pretty large (1337), it is hard to spot your changes and the effect of the merge. Can you refine it to include only small changes? If you've worked on a different branch, then it is best if you make your pull request on that branch, otherwise the changes will be too many to track.

saifk16 commented:
I changed the base branch from `master` to `sv-comp`.
Now my commits and file changes are trackable, earlier I commit to merge the changes to the `master` branch which was commit a lot more changes from the `sv-comp` branch.

@sohah

saifk16 commented:
> @saifk16 , thanks for your contribution. This pull request is pretty large (1337), it is hard to spot your changes and the effect of the merge. Can you refine it to include only small changes? If you've worked on a different branch, then it is best if you make your pull request on that branch, otherwise the changes will be too many to track.

I changed the base branch from `master` to `sv-comp`. Now, my commits and file changes are correctly tracked. Earlier, when I tried to merge my commits into the `master` branch, it included many additional changes from the `sv-comp` branch.

@sohah

saifk16 commented:
### A side-by-side comparison of some test cases I ran:

| Expression                                    | MathMine.max()** Output | Math.max()* Output |
|----------------------------------------------|----------------------|------------------|
| `(NaN, NaN)`                      | NaN                  | NaN              |
| `(NaN, 5.0)`                      | NaN                  | NaN              |
| `(5.0, NaN)`                      | NaN                  | NaN              |
| `(Double.POSITIVE_INFINITY, 1000.0)` | Infinity             | Infinity         |
| `(Double.NEGATIVE_INFINITY, 1000.0)` | 1000.0               | 1000.0           |
| `(-42.0, 42.0)`                   | 42.0                 | 42.0             |
| `(-10.5, -5.5)`                   | -5.5                 | -5.5             |
| `(0.0, -0.0)`                     | 0.0                  | 0.0              |
| `(subnormal1, subnormal2)`        | 1.0E-323             | 1.0E-323         |
| `(5.5, 5.5)`                      | 5.5                  | 5.5              |
| `(Double.MAX_VALUE, -Double.MAX_VALUE)` | 1.7976931348623157E308 | 1.7976931348623157E308 |

> **_Note:_** Changes were made to `max()` and `min()` functions for both `double` and `float` types.

--- 

\** MathMine.max()
 - I created a class `MathMine.java` locally, which mirrors `jpf-symbc/src/classes/java/lang/Math.java`.
  - After successfully running the test cases, I applied the changes to `jpf-symbc/src/classes/java/lang/Math.java`.
  
\* Math.max()
-  The standard Java implementation of `max()` from `java.lang.Math`.

@sohah

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -38,7 +38,7 @@
 
 public class Math {
-	public static final double PI = 3.141592653589793;
+	public static final double PI = 3.14159265358979323846;
```
this will go out of the double representation, that is although your PI value is more accurate, it will be truncated to 3.141592653589793 in java double representation

saifk16 replied:
I searched a little bit, and saw that open jdk follows IEEE 754 standard, I also read the 754-2019 revision of IEEE Standard Floating point (754-2008). Which states for binary64 double precision it would be 1 sign bit, 11 exponent bits, and 52 significand bits (mantissa).

I saw that the jdk8 in open jdk as well as in adoptium temurin they have `3.14159265358979323846`, this value for PI:
[OpenJDK jdk8u-dev](https://github.com/openjdk/jdk8u-dev/blob/master/jdk/src/share/classes/java/lang/Math.java#L123)
[Adoptium jdk8u](https://github.com/adoptium/jdk8u/blob/master/jdk/src/share/classes/java/lang/Math.java)

So as spf works on jdk 8 I thought this would not cause an issue, sorry for the edit ,I will revert back.

This value of PI changed from jdk19
[OpenJDK jdk19u](https://github.com/openjdk/jdk19u/blob/master/src/java.base/share/classes/java/lang/Math.java#L144)

@sohah

sohah replied:
@saifk16 , you are right in referring the openjdk and also the adopitum and how the IEEE 754 standard specifies the precission of doubles. However, using 52-bit for mantissa you can only represent from 15-17 digits. For this particular PI value it will use 15 digits to represent the mantissa, which will result in why it'll be truncated if you add more precision. You can actually see that if you test this in a simple print statement, you'll be able to see the truncation.
It looks like in Java 19 they realized that they do not need the extra precision so they've removed it as you referenced. 
Can you just include a comment about why the choice of the PI representation truncated, so when we go back to this we remember why we took this decision that is different from how PI is in Math.java in java 8?

saifk16 replied:
Added the comment in the Math.java file, please review @sohah

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -58,11 +58,19 @@ public static long abs ( long a) {
 
       //	 TODO:
 	  public static double max ( double a, double b) {
+		if (!(a == a))
+			return a;
+		if (a == b)
+			return (1.0d / a > 0) ? a : b;
```
where did you get this code from? Is there a reference for it? I think it is perhaps better if we use code similar to the one that exists in Math.java in the openjdk

saifk16 replied:
I found out a way to check specifically for **0.0** and **-0.0**. I do take some help from Math.java implementation in OpenJDK, but I thought it would be a complete plagiarism, so I avoided copying it directly from the methods.

To avoid plagiarism, I used **!(a==a)**, this also produces the same output as **a != a**, but is less conventional. Will also revert this. 

> The equality operator == returns false if either operand is NaN.
> In particular, x!=x is true if and only if x is NaN.

In `return (1.0d / a > 0) ? a : b;`  I knew that the default values are 0.0d and 0.0f for double and float respectively, so when **a == b** at the time of **0.0 == -0.0** which would be equal, as we know in java floating-point literals are of double type by default, so I concluded,

```java
1.0d / a > 0
// this would result in +Infinity which is greater than 0
// hence it would return 0.0 i.e., a in this case.
// if values passed are 0.0, -0.0
```

lets say we pass values like this -0.0, 0.0 to max method

```java
1.0d / a > 0
// this would result in -Infinity which is less than 0
// hence again it would return 0.0 i.e, b in this case.
```

I did not understood how raw bits and bit pattern are working so I came up with this code.

Reference => [JLS Numerical Equality Operators == and !=](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.21.1)

Reference => [JLS Floating Point Types and Values](https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.2.3-220)

@sohah

sohah replied:
@saifk16 , I think you might have been on the right track here. Can you please check the math.java class in jpf-core, I think in java11 version they should have a model for it. Can you please check this out, and see to what extend is it different from the variations that you have here? How close is it to the openjdk. Thank you!

saifk16 replied:
Added the comment under the Math.java file, please review @sohah

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -58,11 +58,19 @@ public static long abs ( long a) {
 
 	  // TODO: need to model NaN et al.
 	  public static float max ( float a, float b) {
+		if (!(a == a))
+			return a;
+		if (a == b)
+			return (1.0f / a > 0) ? a : b;
```
same as my comment on `max` methode

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -76,11 +84,19 @@ public static long max ( long a, long b) {
+		if (!(a == a))
+			return a;
+		if (a == b)
+			return (1.0d / a > 0) ? a : b;
```
see my comment on `max` method

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -76,11 +84,19 @@ public static long max ( long a, long b) {
 
 	  // TODO:
 	  public static float min ( float a, float b) {
+		 if (!(a == a))
+			 return a;
+		 if (a == b)
+			 return (1.0f / a > 0) ? a : b;
```
same as my comment on the `max` method

sohah reviewed (changes requested):
Thank you for your contribution, I left you some review comments, can you please get to them. Also, can you please include test cases that show the benefit of the code the you've added. This allows us to validate the code, and to increase our regression tests.

saifk16 commented:
> Thank you for your contribution, I left you some review comments, can you please get to them. Also, can you please include test cases that show the benefit of the code the you've added. This allows us to validate the code, and to increase our regression tests.

Where can I add test cases should I comment them here, should I add them under the test directory 
`jpf-symbc/src/tests/gov/nasa/jpf/symbc`

Or should I add them under the directory where Math.lang is located in jpf-symbc

saifk16 committed:
Reverted PI value change; Updated max/min as per review

sohah commented:
Thank you @saifk16 , please add your test as an example under `src/examples`, you will need a test and a .jpf file. To test your code, you don't really need to have anything symbolic, so you can setup your .jpf to run the test concretely.

f-ei8ht commented:
> Thank you @saifk16 , please add your test as an example under `src/examples`, you will need a test and a .jpf file. To test your code, you don't really need to have anything symbolic, so you can setup your .jpf to run the test concretely.

Okay, Thanks!!!

saifk16 commented:
> Thank you @saifk16 , please add your test as an example under `src/examples`, you will need a test and a .jpf file. To test your code, you don't really need to have anything symbolic, so you can setup your .jpf to run the test concretely.

Okay ma'am, Thanks.

saifk16 committed:
Add test cases for Math.max() and Math.min()

saifk16 commented:
> Thank you @saifk16 , please add your test as an example under `src/examples`, you will need a test and a .jpf file. To test your code, you don't really need to have anything symbolic, so you can setup your .jpf to run the test concretely.

Added test cases that will show the benefits of my code I had added, will also help in validation.

In my first commit, I also saw that the `double` type implementation of `Math.min()` was a bit wrong.
```java
public static double min ( double a, double b) {
	return (a >= b) ? a : b;
}
``` 

It should be `a <= b`, so I forgot to mention this in my first commit.

@sohah ma'am please review my test cases.

Thanks.

sohah commented on jpf-symbc/src/examples/TestMathMaxMin.java:
```diff
@@ -0,0 +1,106 @@
+
+    public static void testFloatMax() {
+
+        // Tests similar to the cases above.
+        System.out.println(Math.max(0.0f, -0.0f)); // 0.0
+        System.out.println(Math.max(-0.0f, 0.0f)); // 0.0
+        System.out.println(Math.max(Float.NaN, 4.5)); // NaN
+        System.out.println(Math.max(4.5, Float.NaN)); // NaN
```
Can you please make these a set of assert statements as you've done with testDouble Max?

sohah commented on jpf-symbc/src/examples/TestMathMaxMin.java:
```diff
@@ -0,0 +1,106 @@
+     * This method tests {@code Math.min()} for {@code double} type values.
+     */
+
+    public static void testDoubleMin() {
+
+        // Earlier this was returning 0.0, -0.0, 4.5, NaN and 9.0
+        System.out.println(Math.min(0.0, -0.0)); // 0.0
+        System.out.println(Math.min(-0.0, 0.0)); // 0.0
+        System.out.println(Math.min(Double.NaN, 4.5)); // NaN
+        System.out.println(Math.min(4.5, Double.NaN)); // NaN
+        System.out.println(Math.min(5.0, 9.0)); // 5.0
```
Can you please make these a set of assert statements as you've done with testDouble Max?

sohah commented on jpf-symbc/src/examples/TestMathMaxMin.java:
```diff
@@ -0,0 +1,106 @@
+     * This method tests {@code Math.min()} for {@code double} type values.
+     */
+
+    public static void testFloatMin() {
+
+        // Tests similar to the cases above.
+        System.out.println(Math.min(0.0f, -0.0f)); // 0.0
+        System.out.println(Math.min(-0.0f, 0.0f)); // 0.0
+        System.out.println(Math.min(Float.NaN, 4.5)); // NaN
+        System.out.println(Math.min(4.5, Float.NaN)); // NaN
```
Can you please make these a set of assert statements as you've done with testDouble Max?

sohah reviewed (changes requested):
@saifk16 , thank you so much for your contributions. Just a few more comments before we get this pr in.

saifk16 commented:
> @saifk16 , thank you so much for your contributions. Just a few more comments before we get this pr in.

Refactored min logic and updated the test cases, now they use the assert statement.
Also, by mistake I committed from a different account of mine, sorry for that.

`Math.min()` should use `a < 0` instead of `a > 0` as in `double max` and `float max`.

Next up:
- including the comment for PI
- math.java class in jpf-core

@sohah

saifk16 committed:
JavaDoc comment for the PI constant

saifk16 commented:
Added a comment explaining the choice of the PI constant and the truncation due to double precision limits.

Tried renaming the source branch, but since GitHub doesn't allow updating the source branch of an open PR, I had to re-commit the last changes along with the new ones.

This follows from the investigation and discussion here: [comment](https://github.com/SymbolicPathFinder/jpf-symbc/pull/111#discussion_r2022622500)

For future work on `Math.rint` and other methods in `Math.java`, I'm thinking of moving that to a new branch and creating a separate PR. This PR can then be referenced in that PR for the work that is done till now.

@sohah

sohah commented:
> > @saifk16 , thank you so much for your contributions. Just a few more comments before we get this pr in.
> 
> Refactored min logic and updated the test cases, now they use the assert statement. Also, by mistake I committed from a different account of mine, sorry for that.
> 
> `Math.min()` should use `a < 0` instead of `a > 0` as in `double max` and `float max`.
> 
> Next up:
> 
> * including the comment for PI
> * math.java class in jpf-core
> 
> @sohah

Hi @saifk16 , have you looked at the math.java class in jpf-core in java-11? it might be better to use their models in the version that we have here

f-ei8ht commented:
> > > @saifk16 , thank you so much for your contributions. Just a few more comments before we get this pr in.
> > 
> > 
> > Refactored min logic and updated the test cases, now they use the assert statement. Also, by mistake I committed from a different account of mine, sorry for that.
> > `Math.min()` should use `a < 0` instead of `a > 0` as in `double max` and `float max`.
> > Next up:
> > 
> > * including the comment for PI
> > * math.java class in jpf-core
> > 
> > @sohah
> 
> Hi @saifk16 , have you looked at the math.java class in jpf-core in java-11? it might be better to use their models in the version that we have here

Hello @sohah ma'am are you talking about this => [file](https://github.com/javapathfinder/jpf-core/blob/70c3013c60d627bc8158316abcf82fe503f5ae15/src/peers/gov/nasa/jpf/vm/JPF_java_lang_Math.java)

This has in the native peers concept, but I cannot find any branch related to java 11 there is java 8 java 12 and java 17 and master which one is related to java 11?

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -58,11 +74,19 @@ public static long abs ( long a) {
 
       //	 TODO:
```
can you please remove this

saifk16 replied:
Okay ma'am done.

sohah commented on jpf-symbc/src/classes/java/lang/Math.java:
```diff
@@ -58,11 +74,19 @@ public static long abs ( long a) {
 		return (a >= b) ? a : b;
 	  }
 
 	  // TODO: need to model NaN et al.
```
Since you've modeled NaN can you please this comment and similar ones in all other places?

saifk16 replied:
Removed the comment, need to add ? some comment or just leave it like that ?

sohah reviewed (changes requested):
@saifk16 Thank you for your contribution, I think we can accept this, but please clean up the code from unnecessary comments first, and then we can include this code in. Thank you.

saifk16 committed:
Removed unnecessary comments

sohah commented:
@saifk16 , thank you for your changes.
@yannicnoller , can you please accept this pull request.

yannicnoller commented:
Thanks @sohah and @saifk16! Merged!

saifk16 commented:
> Thanks @sohah and @saifk16! Merged!

Thanks @yannicnoller sir for merging
